modules/extension/wmts/src/main/java/org/geotools/ows/wmts/map/WMTSCoverageReader.java

```diff
@@ -563,4 +564,141 @@ public String getRequestedTime() {
+                    coverage2D,
+                    destEnvelope.getCoordinateReferenceSystem(),
+                    interpolation,
+                    destEnvelope,
+                    readParameters.getBKGArray(),
+                    factory,
+                    newHints);
+        } catch (FactoryException factoryException) {
+            if (LOGGER.isLoggable(Level.SEVERE))
+                LOGGER.log(
+                        Level.SEVERE,
+                        errorMessage(destEnvelope.getCoordinateReferenceSystem()),
+                        factoryException);
+            throw new RuntimeException(factoryException);
+        }
+    }
+
+    private String errorMessage(CoordinateReferenceSystem targetCrs) {
+        StringBuilder msg =
+                new StringBuilder("Something wrong happended while trying to reproject ");
+        if (coverageName != null) msg.append(coverageName).append(" ");
+        msg.append("WMTS coverage to ");
+        if (crs != null) msg.append(CRS.toSRS(targetCrs));
+        else msg.append(" null srs.");
+        return msg.toString();
+    }
+
+    private GridCoverage2D getMapReproject(
+            ReferencedEnvelope requestedEnvelope, String time, WMTSReadParameters readParameters)
+            throws IOException {
+        try {
+            CoordinateReferenceSystem targetCRS = requestedEnvelope.getCoordinateReferenceSystem();
+            ReferencedEnvelope nativeEnvelope =
+                    requestedEnvelope.transform(getBestSourceCRS(targetCRS, readParameters), false);
+            ReferencedEnvelope gridEnvelope =
+                    initTileRequest(
+                            nativeEnvelope,
+                            readParameters.getWidth(),
+                            readParameters.getHeight(),
+                            time);
+            GridCoverage2D result = getMap(nativeEnvelope, gridEnvelope);
+            if (!CRS.equalsIgnoreMetadata(result.getCoordinateReferenceSystem(), targetCRS))
+                result =
+                        reprojectManually(
+                                result, new GeneralEnvelope(requestedEnvelope), readParameters);
+            return result;
+        } catch (FactoryException | TransformException e) {
+            if (LOGGER.isLoggable(Level.SEVERE))
+                LOGGER.log(
+                        Level.SEVERE,
+                        "Error while reprojecting the requested envelope to the selected native crs.",
+                        e);
+            throw new IOException(e);
+        }
+    }
+
+    // get the CRS to be used to reproject to a CRS that the server doesn't support.
+    CoordinateReferenceSystem getBestSourceCRS(
+            CoordinateReferenceSystem targetCRS, WMTSReadParameters readParameters)
+            throws IOException {
+        try {
+            CoordinateReferenceSystem result = null;
+            String switched = switchDefinition(targetCRS);
```

Comment:
I am confused... why try directly to switch to another style? Ah because by the time you call this method, you already know the targetCRS identifier is not among the supported ones?

Reply:
yes the getBestSourceCRS should be called when we know the targetCRS is not among the supported ones. I've added a comment.